parser.py

```diff
@@ -4,42 +4,132 @@
 from utils import *
 from lxml import etree
 
-class State:
-    class Waiting: pass
-    class Meta: pass
-    class Ingredients: pass
-    class Steps: pass
+
+class Line:
+    """Holds the type and contents of a line, e.g. an ingredient or an instruction.
+
+    In case that word wrapping was used in the original markup, one Line class can
+    actually hold the contents of multiple lines in the source code belonging to the
+    same ingredient, instruction etc.
+    """
+    class Empty: pass
+    class Plain: pass
+    class Comment: pass
+    class Command: pass
+    class Part: pass
+    class Ingredient: pass
+    class Step: pass
+    class Note: pass
     class WaitPhase: pass
-    class FinishPhase: pass
-    class FinishRecipe: pass
 
-def parseIngredient(line):
-    r = r'^\s*#\s*([0-9/\.]+(?:\s+[0-9/]+)?)([a-zA-Z]+)?\s+(.*)'
-    result = re.match( r, line)
+    startCharacters = (
+        ("'", Comment),
+        ('!', Command),
+        ('#', Part),
+        ('*', Ingredient),
+        ('--', Note), # notes have to be parsed before steps!
+        ('-', Step),
+        ('+', WaitPhase),
+    )
+
+    def __init__(self, lineType = None, lineNo = None, contents = None):
+        self.lineType = lineType
+        self.lineNo = lineNo
+        self.contents = contents
+
+    def __repr__(self):
+        return 'Line({!r}, {!r}, {!r})'.format(self.lineType, self.lineNo, self.contents)
+
+    @property
+    def contents(self):
+        if isinstance(self._contents, str):
+            return self._contents
+        else:
+            return ' '.join(self._contents)
 
-    if result is None:
-        line = line.lstrip('# \t')
-        line = line.rstrip(' \t\n')
-        return Ingredient( line, None, None)
+    @contents.setter
+    def contents(self, contents):
+        self._contents = contents
+
+    def append(self, contents):
+        if isinstance(self._contents, str):
+            self._contents = [self._contents, contents]
+        else:
+            self._contents.append(contents)
+
+    def __eq__(self, other):
+        if not isinstance(other, Line):
+            return NotImplemented
+
+        return (self.lineType == other.lineType
+            and self.lineNo == other.lineNo
+            and self.contents == other.contents)
+
+    def __ne__(self, other):
+        return not self == other
+
+def preprocessLines(stream):
+    """Generator that joins word wrapped lines back together and parses the type of the line,
+    e.g. ingredient or step. Yields instances of the Line class.
+    """
+    currentLine = None
+    for lineNo, rawLine in enumerate(stream, 1):
+        lineType = None
+        contents = rawLine.strip()
+        for c, t in Line.startCharacters:
```

Comment:
It needs to be ordered since '--' has to be checked before '-'. So if I would want to use a dict, it would have to be an OrderedDict (which is pointless since the only thing I do with it is to iterate over it).